packages/openid4vc-client/src/OpenId4VcClientService.ts

```diff
@@ -1,10 +1,201 @@
+      const key = getKeyFromVerificationMethod(verificationMethod)
+
+      const payload = JsonEncoder.toBuffer(jwt.payload)
+
+      const jws = await this.jwsService.createJwsCompact(agentContext, {
+        key, // FIXME null check
+        payload,
+        protectedHeaderOptions: {
+          alg: jwt.header.alg,
+          kid: jwt.header.kid,
+        },
+      })
+
+      return jws
+    }
+  }
+
+  private getSignCallback(agentContext: AgentContext) {
+    return {
+      signCallback: this.signCallback(agentContext),
+    }
+  }
+
+  public async preAuthorized(
+    agentContext: AgentContext,
+    options: PreAuthorizedOptions,
+    checkRevocationState = true
+  ): Promise<W3cCredentialRecord> {
+    this.logger.debug('Running pre-authorized flow with options', options)
+
+    // The clientId is set to a dummy value for now as we don't have
+    // one in the pre-auth flow, but it's currently required by OpenID4VCIClient.
+    const clientId = 'some-client'
+
+    const client = await OpenID4VCIClient.initiateFromURI({
+      issuanceInitiationURI: options.issuerUri,
+      flowType: AuthzFlowType.PRE_AUTHORIZED_CODE_FLOW,
+      kid: options.kid,
+      alg: Alg.EdDSA,
+      clientId: clientId,
+    })
+
+    const accessToken = await client.acquireAccessToken({ clientId: clientId })
+
+    this.logger.info('Fetched server accessToken', accessToken)
+
+    // We currently need the ts-ignore because the type
+    // inside of OpenID4VCIClient needs to be updated.
+    // @ts-ignore
+    if (!accessToken.scope) {
+      throw new AriesFrameworkError(
+        "Access token response doesn't contain a scope. Only scoped issuer URIs are supported at this time."
+      )
+    }
+
+    const serverMetadata = await client.retrieveServerMetadata()
+
+    this.logger.info('Fetched server metadata', {
+      issuer: serverMetadata.issuer,
+      credentialEndpoint: serverMetadata.credential_endpoint,
+      tokenEndpoint: serverMetadata.token_endpoint,
+    })
+
+    this.logger.debug('Full server metadata', serverMetadata)
+
+    // proof of possession
+    const callbacks = this.getSignCallback(agentContext)
+
+    const proofInput = await ProofOfPossessionBuilder.fromAccessTokenResponse({
+      accessTokenResponse: accessToken,
+      callbacks: callbacks,
+    })
+      .withEndpointMetadata(serverMetadata)
+      .withClientId(clientId)
+      .withAlg(Alg.EdDSA)
+      .withKid(options.kid)
+      .build()
+
+    this.logger.debug('Generated JWS', proofInput)
+
+    const credentialRequestClient = CredentialRequestClientBuilder.fromIssuanceInitiationURI({
+      uri: options.issuerUri,
+      metadata: serverMetadata,
+    })
+      .withTokenFromResponse(accessToken)
+      .build()
+
+    const credentialResponse = await credentialRequestClient.acquireCredentialsUsingProof({
+      proofInput,
+      // @ts-ignore
+      credentialType: accessToken.scope,
+      format: 'ldp_vc', // Allows us to override the format
```

Comment:
should we check though if the `foramt` is present in the issuer metadata, or will that be handled by the sphereon lib?